specs/simpleserialize.md

```diff
@@ -0,0 +1,230 @@
+| `len`        | Length/Number of Bytes.                                                                        |
+| `to_bytes`   | Convert to bytes. Should take parameters ``size`` and ``byte_order``.                          |
+| `from_bytes` | Convert form bytes to object. Should take ``bytes`` and ``byte_order``.                        |
+| `value`      | The value to serialize.                                                                        |
+| `rawbytes`   | Raw serialized bytes.                                                                          |
+
+## Constants
+
+| Constant       | Value | Definition                                                              |
+|:---------------|:-----:|:------------------------------------------------------------------------|
+| `LENGTH_BYTES` |   4   | Number of bytes used for the length added before the serialized object. |
+
+
+## Overview
+
+### Serialize/Encode
+
+#### uint: 8/16/24/32/64/256
+
+Convert directly to bytes the size of the int. (e.g. ``uint16 = 2 bytes``)
+
+All integers are serialized as **big endian**.
+
+| Check to perform                 | Code                    |
+|:---------------------------------|:------------------------|
+| Size is a byte integer           | ``int_size % 8 == 0``   |
+| Value is less than max           | ``2**int_size > value`` |
+
+```python
+buffer_size = int_size / 8
+return value.to_bytes(buffer_size, 'big')
+```
+
+#### Address
+
+The address should already come as a hash/byte format. Ensure that length is
+**20**.
+
+| Check to perform       | Code                 |
+|:-----------------------|:---------------------|
+| Length is correct (20) | ``len(value) == 20`` |
+
+```python
+assert( len(value) == 20 )
+return value
+```
+
+#### Hash32
+
+The hash32 should already be a 32 byte length serialized data format. The safety
+check ensures the 32 byte length is satisfied.
+
+| Check to perform       | Code                 |
+|:-----------------------|:---------------------|
+| Length is correct (32) | ``len(value) == 32`` |
+
+```python
+assert( len(value) == 32 )
+return value
+```
+
+#### Bytes
+
+For general `byte` type:
+1. Get the length/number of bytes; Encode into a 4 byte integer.
+2. Append the value to the length and return: ``[ length_bytes ] + [
+   value_bytes ]``
+
+| Check to perform                     | Code                   |
+|:-------------------------------------|:-----------------------|
+| Length of bytes can fit into 4 bytes | ``len(value) < 2**32`` |
+
+```python
+byte_length = (len(value)).to_bytes(LENGTH_BYTES, 'big')
+return byte_length + value
+```
+
+#### List
+
+For lists of values, get the length of the list and then serialize the value
+of each item in the list:
+1. For each item in list:
+   1. serialize.
+   2. append to string.
+2. Get size of serialized string. Encode into a 4 byte integer.
```

Comment:
This part is Python specific. This should read as

```Markdown
#### List/Vectors

1. Get the number of raw bytes to serialize: it is `len(list) * sizeof(element)`.
    Encode that as a 4-byte bigEndian uint32
2. Append your elements in a packed manner.

* Note on efficiency: consider using a container that does not need to iterate over all elements to get its length. For example Python lists, C++ vectors or Rust Vec.
```

Then your Python implementation notes.

The perf notes are to address the problem highlighted by @AlexeyAkhunov, if the container doe keep track of its length (for example LinkedLists) prefixing the length is inefficient.

Reply:
Thanks! This is great. 

I think I'll add yours specifically, since it's more generic and should cater to larger of an audience! 

Thanks for pointing me to the notes, I'll make sure to run through them and give a reference! :smile:

Reply:
You can refer to this post: https://github.com/ethereum/eth2.0-specs/issues/2#issuecomment-425626242

> ## On length prefix vs length suffix
>
> https://github.com/ethereum/beacon_chain/issues/94 and https://github.com/ethereum/eth2.0-pm/issues/8
>
> - The argument for moving at the end is because otherwise you would have to iterate on the data structure once to get the length then once again to get feed the data. ==> All languages even Haskell have a vector implementation that tracks length so you wouldn't need the first iteration to get length.
>
> - Furthermore putting length at the end means you need specific token to indicate end of data like ‘\0’ in cstring. This is causing a lot of issue in C code also this prevents reading by chunk into a buffer, you would have to read token by token (or read by chunk and backtrack).

Reply:
Thanks @mratsim! :+1: 

I'll hold this PR open until we get more clarification/acceptance?

Reply:
I definitely don't want to block it, just thought it will be nice to have more eyes on it, like @mratsim. 👍👀 It looks good for me to merge it, as a clear document to be audited and to be disucussed.